### PR TITLE
ARROW-16741: [C++] Add Benchmarks for Binary Temporal Operations

### PR DESCRIPTION
**Pull request by iChauster**
Add all binary temporal benchmarks and documentation to `api_scalar.h`

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (rok) [inline cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:243]: Perhaps we could also add `add`,  `add_checked`, `subtract`, `subtract_checked`, `multiply`, `multiply_checked`, `divide` and `divide_checked` here since they are binary?
These kernels are [defined here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L1999). However they are a bit more heterogeneous type wise so they will require some care (e.g. `add(timestamp, timestamp)` is invalid, while `subtract(timestamp, timestamp)` isn't)

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (rok) [inline cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:199]: For some cases you might want other time types here (date32 date64, time32, time64, duration).

File: cpp/src/arrow/compute/api_scalar.h
Comment (rok) [inline cpp/src/arrow/compute/api_scalar.h:1604]: @lidavidm was this not mapped out for a reason?

**Review by rok — COMMENTED** [review]
Thanks for doing this. It is very much needed!

File: cpp/src/arrow/compute/api_scalar.h
Comment (lidavidm) [reply cpp/src/arrow/compute/api_scalar.h:1604]: I don't think there was any reason, I just neglected to add those.

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (iChauster) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:199]: Hi @rok, thanks for the timely review! 

I was a bit curious about this, since it seems for `BenchmarkTemporalBinary` and `BenchmarkTemporal` (the original unary version), we are picking random `int64_t` -- which I assume is `date64`. How do we achieve the other time types, and is there some helpful example code I can follow?

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (iChauster) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:243]: They are binary, but I believe these kernels are already benchmarked in `scalar_arithmetic_benchmark.cc`. Does this mean adding support for adding, subtracting, (multiplying and dividing!) timestamps, and then benchmarking those implementations?

**Comment by iChauster** [comment]
Hi @rok, unfortunately I can't re-request review as a first-time contributor but I was wondering if you could answer some of my clarification questions from a few days ago when you get the chance! Excited to get this merged in.

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:243]: `scalar_arithmetic_benchmark.cc` benchmarks non-temporal arrays right now. It would be useful to also benchmark for temporal zoned/nonzoned temporal types as computation would follow a different codepath for those.

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:199]: [A more complete example.](https://github.com/apache/arrow/blob/master/cpp/src/arrow/testing/generator.cc#L101-L155)

**Comment by rok** [comment]
Hey @iChauster sorry for the delay!

File: cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc
Comment (iChauster) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:199]: Hi @rok, thank you for the clarification!

I refactored the code to work for the other data types, however it seems like these temporal binary functions [currently do not support any other datatype](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_temporal_binary.cc#L461) than `Int64`.

When running my code with something like `date32` or `time64` with `random.ArrayOf`, I get an `NotImplemented: Function years_between has no kernel matching input types (array[time64[ns]], array[time64[ns]])`

Comment (iChauster) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc:199]: Not sure why, but it seems like `date64` is viable as an input type, but `time64` / `time32`, and `duration` are not.

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc]: Yeah, not all kernels will support all types.

**Comment by rok** [comment]
> @rok, excellent, I think the binary benchmarks now support different datatypes for their respective kernels. Is this ready to be merged now?

Looks good to me. Do benchmarks run ok for you locally?

We need a committer to merge this. @lidavidm probably knows these functions best.

**Review by lidavidm — APPROVED** [review]

**Comment by iChauster** [comment]
@rok the benchmarks run great locally and fall into a similar range as the `TemporalRounding` statistics.

Thanks @lidavidm for approving, I'm assuming someone with write access will merge eventually?

**Comment by lidavidm** [comment]
I have write access yes was just waiting for CI to pass/it was the end of the day for me :)

**Comment by rok** [comment]
@iChauster here are the [temporal benchmarks for this PR](https://conbench.ursa.dev/compare/runs/85289eeb3c794766948e9065016c4717...aabd4c1c4d934cfa81256b6f772453f1/) if you're curious.

**Pull request merged**
